File: node/src/resources/fleets.ts
/**
 * `client.fleets` — list / get / create / update / disable, plus `members` and
 * `machines` sub-resources (lazy getters). `get` scans `list` client-side;
 * `disable` is soft-disable.
 */

import { MfaRequirement } from "../_proto/cmdop/core/v1/common_pb";
import type {
  FleetList,
  FleetSummary,
  MachineFleetLink,
  MemberInfo,
  MemberList,
} from "../_proto/cmdop/core/v1/fleets_pb";
import type { MachineList, MachineSummary } from "../_proto/cmdop/core/v1/machines_pb";
import { BaseResource, type ListPageOptions } from "./base";
import type { ListMachinesOptions } from "./machines";

const MFA: Record<string, MfaRequirement> = {
  none: MfaRequirement.NONE,
  optional: MfaRequirement.OPTIONAL,
  required: MfaRequirement.REQUIRED,
};

export class FleetsResource extends BaseResource {
  #members?: FleetMembersResource;
  get members(): FleetMembersResource {
    return (this.#members ??= new FleetMembersResource(this.client));
  }

  #machines?: FleetMachinesResource;
  get machines(): FleetMachinesResource {
    return (this.#machines ??= new FleetMachinesResource(this.client));
  }

  async list(opts: ListPageOptions = {}): Promise<FleetList> {
    const env = await this.unary(
      this.req({ case: "listFleetsReq", value: { page: this.offsetPage(opts.page, opts.perPage) } }),
    );
    return env.payload.value as FleetList;
  }

  iter(opts: ListPageOptions = {}): AsyncGenerator<FleetSummary> {
    return this.paginateOffset((page) => this.list({ page, perPage: opts.perPage }));
  }

  pages(opts: ListPageOptions = {}): AsyncGenerator<FleetList> {
    return this.pagesOffset((page) => this.list({ page, perPage: opts.perPage }));
  }

  /** Get a single fleet by scanning the (offset-paginated) list. */
  async get(fleetId: string): Promise<FleetSummary> {
    for await (const f of this.iter()) {
      if (f.id === fleetId) return f;
    }
    const { NotFoundError } = await import("../errors");
    throw new NotFoundError(`Fleet ${fleetId} not found`, undefined, "not_found");
  }

  async create(body: {
    name: string;
    slug: string;
    mfaRequired?: string;
  }): Promise<FleetSummary> {
    const env = await this.unary(
      this.req({
        case: "createFleetReq",
        value: {
          name: body.name,
          slug: body.slug,
          mfaRequired: MFA[body.mfaRequired ?? "none"] ?? MfaRequirement.NONE,
        },
      }),
    );
    return env.payload.value as FleetSummary;
  }

  async update(
    fleetId: string,
    body: { name?: string; mfaRequired?: string } = {},
  ): Promise<FleetSummary> {
    const value: { fleetId: string; name: string; mfaRequired?: MfaRequirement } = {
      fleetId,
      name: body.name ?? "",
    };
    if (body.mfaRequired !== undefined) {
      value.mfaRequired = MFA[body.mfaRequired] ?? MfaRequirement.NONE;
    }
    const env = await this.unary(this.req({ case: "updateFleetReq", value }));
    return env.payload.value as FleetSummary;
  }

  /** Soft-disable (not delete). */
  async disable(fleetId: string): Promise<void> {
    await this.unary(this.req({ case: "disableFleetReq", value: { fleetId } }));
  }
}

export class FleetMembersResource extends BaseResource {
  async list(fleetId?: string, opts: ListPageOptions = {}): Promise<MemberList> {
    const fid = this.fleet(fleetId);
    const env = await this.unary(
      this.req({
        case: "listMembersReq",
        value: { fleetId: fid, page: this.offsetPage(opts.page, opts.perPage) },
      }),
    );
    return env.payload.value as MemberList;
  }

  iter(fleetId?: string, opts: ListPageOptions = {}): AsyncGenerator<MemberInfo> {
    return this.paginateOffset((page) => this.list(fleetId, { page, perPage: opts.perPage }));
  }

  async add(
    body: { email: string; role?: string },
    opts: { fleetId?: string } = {},
  ): Promise<MemberInfo> {
    const fid = this.fleet(opts.fleetId);
    const env = await this.unary(
      this.req({ case: "addMemberReq", value: { fleetId: fid, email: body.email, role: body.role ?? "member" } }),
    );
    return env.payload.value as MemberInfo;
  }

  async setRole(userId: string, role: string, opts: { fleetId?: string } = {}): Promise<MemberInfo> {
    const fid = this.fleet(opts.fleetId);
    const env = await this.unary(
      this.req({ case: "setMemberRoleReq", value: { fleetId: fid, userId, role } }),
    );
    return env.payload.value as MemberInfo;
  }

  async remove(userId: string, opts: { fleetId?: string } = {}): Promise<void> {
    const fid = this.fleet(opts.fleetId);
    await this.unary(this.req({ case: "removeMemberReq", value: { fleetId: fid, userId } }));
  }
}

export class FleetMachinesResource extends BaseResource {
  async list(fleetId?: string, opts: ListMachinesOptions = {}): Promise<MachineList> {
    const fid = this.fleet(fleetId);
    const env = await this.unary(
      this.req({
        case: "listFleetMachinesReq",
        value: {
          fleetId: fid,
          presence: opts.presence ?? "any",
          q: opts.q ?? "",
          limit: opts.limit ?? 100,
          cursor: opts.cursor ?? "",
        },
      }),
    );
    return env.payload.value as MachineList;
  }

  iter(fleetId?: string, opts: ListMachinesOptions = {}): AsyncGenerator<MachineSummary> {
    return this.paginateCursor((cursor) => this.list(fleetId, { ...opts, cursor }));
  }

  async attach(machineId: string, opts: { fleetId?: string } = {}): Promise<MachineFleetLink> {
    const fid = this.fleet(opts.fleetId);
    const env = await this.unary(
      this.req({ case: "attachMachineReq", value: { fleetId: fid, machineId } }),
    );
    return env.payload.value as MachineFleetLink;
  }

  async detach(machineId: string, opts: { fleetId?: string } = {}): Promise<void> {
    const fid = this.fleet(opts.fleetId);
    await this.unary(this.req({ case: "detachMachineReq", value: { fleetId: fid, machineId } }));
  }
}

File: python/src/cmdop/resources/fleets.py
"""fleets namespace + fleets.members + fleets.machines sub-resources.

``get`` scans ``list`` client-side (no dedicated op); ``disable`` is soft-disable.
``members`` / ``machines`` are fleet-scoped (default-fleet fallback).
"""

from __future__ import annotations

from functools import cached_property
from typing import TYPE_CHECKING, Any

from cmdop._proto.cmdop.core.v1 import common_pb2 as common_pb
from cmdop._proto.cmdop.core.v1 import envelope_pb2 as pb
from cmdop._proto.cmdop.core.v1 import fleets_pb2 as f_pb
from cmdop.resources.base import BaseResource

if TYPE_CHECKING:
    from collections.abc import AsyncIterator

    from cmdop.types import (
        FleetList,
        FleetSummary,
        MachineFleetLink,
        MachineList,
        MemberInfo,
        MemberList,
    )

_MFA = {
    "none": common_pb.MFA_REQUIREMENT_NONE,
    "optional": common_pb.MFA_REQUIREMENT_OPTIONAL,
    "required": common_pb.MFA_REQUIREMENT_REQUIRED,
}


class FleetsResource(BaseResource):
    @cached_property
    def members(self) -> FleetMembersResource:
        return FleetMembersResource(self._client)

    @cached_property
    def machines(self) -> FleetMachinesResource:
        return FleetMachinesResource(self._client)

    async def list(self, *, page: int = 1, per_page: int | None = None) -> FleetList:
        req = pb.Envelope(
            list_fleets_req=f_pb.ListFleetsRequest(page=self._offset_page(page, per_page))
        )
        return (await self._unary(req)).fleet_list

    async def get(self, fleet_id: str) -> FleetSummary:
        """Find a fleet by id within the (offset-paginated) list."""
        async for fleet in self.iter():
            if str(fleet.id) == str(fleet_id):
                return fleet
        from cmdop.errors import NotFoundError

        raise NotFoundError(f"Fleet {fleet_id} not found", code="not_found")

    async def create(self, *, name: str, slug: str, mfa_required: str = "none") -> FleetSummary:
        req = pb.Envelope(
            create_fleet_req=f_pb.CreateFleetRequest(
                name=name, slug=slug, mfa_required=_MFA.get(mfa_required, common_pb.MFA_REQUIREMENT_NONE)
            )
        )
        return (await self._unary(req)).fleet_summary

    async def update(
        self, fleet_id: str, *, name: str | None = None, mfa_required: str | None = None
    ) -> FleetSummary:
        body = f_pb.UpdateFleetRequest(fleet_id=str(fleet_id), name=name or "")
        if mfa_required is not None:
            body.mfa_required = _MFA.get(mfa_required, common_pb.MFA_REQUIREMENT_NONE)
        return (await self._unary(pb.Envelope(update_fleet_req=body))).fleet_summary

    async def disable(self, fleet_id: str) -> None:
        """Soft-disable (not delete)."""
        req = pb.Envelope(disable_fleet_req=f_pb.DisableFleetRequest(fleet_id=str(fleet_id)))
        await self._unary(req)

    # -- pagination --------------------------------------------------------

    async def iter(self, *, per_page: int | None = None) -> AsyncIterator[Any]:
        async for page in self.pages(per_page=per_page):
            for item in page.items:
                yield item

    async def pages(self, *, per_page: int | None = None) -> AsyncIterator[FleetList]:
        async def fetch(page_number: int) -> FleetList:
            return await self.list(page=page_number, per_page=per_page)

        async for page in self._paginate_offset(fetch):
            yield page


class FleetMembersResource(BaseResource):
    async def list(
        self, fleet_id: str | None = None, *, page: int = 1, per_page: int | None = None
    ) -> MemberList:
        fid = self._fleet(fleet_id)
        req = pb.Envelope(
            list_members_req=f_pb.ListMembersRequest(
                fleet_id=fid, page=self._offset_page(page, per_page)
            )
        )
        return (await self._unary(req)).member_list

    async def add(self, email: str, role: str = "member", *, fleet_id: str | None = None) -> MemberInfo:
        fid = self._fleet(fleet_id)
        req = pb.Envelope(add_member_req=f_pb.AddMemberRequest(fleet_id=fid, email=email, role=role))
        return (await self._unary(req)).member_info

    async def set_role(self, user_id: str, role: str, *, fleet_id: str | None = None) -> MemberInfo:
        fid = self._fleet(fleet_id)
        req = pb.Envelope(
            set_member_role_req=f_pb.SetMemberRoleRequest(
                fleet_id=fid, user_id=str(user_id), role=role
            )
        )
        return (await self._unary(req)).member_info

    async def remove(self, user_id: str, *, fleet_id: str | None = None) -> None:
        fid = self._fleet(fleet_id)
        req = pb.Envelope(
            remove_member_req=f_pb.RemoveMemberRequest(fleet_id=fid, user_id=str(user_id))
        )
        await self._unary(req)


class FleetMachinesResource(BaseResource):
    async def list(
        self,
        fleet_id: str | None = None,
        *,
        presence: str = "any",
        q: str | None = None,
        limit: int = 100,
        cursor: str | None = None,
    ) -> MachineList:
        fid = self._fleet(fleet_id)
        req = pb.Envelope(
            list_fleet_machines_req=f_pb.ListFleetMachinesRequest(
                fleet_id=fid, presence=presence, q=q or "", limit=limit, cursor=cursor or ""
            )
        )
        return (await self._unary(req)).machine_list

    async def attach(self, machine_id: str, *, fleet_id: str | None = None) -> MachineFleetLink:
        fid = self._fleet(fleet_id)
        req = pb.Envelope(
            attach_machine_req=f_pb.AttachMachineRequest(fleet_id=fid, machine_id=str(machine_id))
        )
        return (await self._unary(req)).machine_fleet_link

    async def detach(self, machine_id: str, *, fleet_id: str | None = None) -> None:
        fid = self._fleet(fleet_id)
        req = pb.Envelope(
            detach_machine_req=f_pb.DetachMachineRequest(fleet_id=fid, machine_id=str(machine_id))
        )
        await self._unary(req)
